### ISSUE TITLE
How to handle gradient accumulation with multiple models ?

### ISSUE THREAD

**Issue by patil-suraj** [huggingface/accelerate#668]
To do gradient accumulation with `accelerate` we wrap the model in `accelerator.accumulate` context.  But what would be the right way to achieve this when multiple models are  involved ?

For example, when training latent diffusion models we have 3 separate models, a vae, text encoder and a unet, as you can see in this [script](https://github.com/huggingface/diffusers/blob/4436d5ff4b2a744eab21d0633d864de2db8734c2/examples/training/textual_inversion.py).  Of which only the text_encoder is being trained (but could also train others as well).

The obvious way to do this would be to create a wrapper model, but curious to know if this can be achieved without using the wrapper model.

cc @muellerzr

**Comment by Lime-Cakes** [comment]
> Re-opening this issue again. For doing grad accum with `accelerator.accumulate` with two models (both are being trained) can we use two context managers like this
> 
> ```python
> with accumulate(model1) as _, with accumulate(model2) as _:
>      training_step()
> ```

Does this currently work? Or is that a feature request, meaning it currently wouldn't work, but would work in the future? Sorry, I got a bit confused by the feature request tag and your comment.

**Comment by pfeatherstone** [comment]
Very interested in this. I'm training two models at once and can only use batch sizes of less than 5 on my machine... So gradient accumulation would be great

**Comment by LvanderGoten** [comment]
I "solved" it by creating one `Accelerator` per model. If you use only one and register the models via `accelerator.prepare(model1, ..., modelN)` at least one of the models is not learning anything. This might be a bug.

**Comment by meng-wenlong** [comment]
As I think, in this case writing accumulation by yourself maybe more flexible. `Accelerator.accumulate()` is not necessary. Just write code like:
```Python
loss = loss / gradient_accumulation_steps
accelerator.backward(loss)
if (index+1) % gradient_accumulation_steps == 0:
    optimizer.step()
    scheduler.step()
    optimizer.zero_grad()
```

**Comment by PrinceRay7** [comment]
This is pretty simple actually, just use "with accelerator.accumulate(model1), accelerator.accumulate(model2): " this is the mechanism of "with", the following code will be in this two contexts, so just simply put them together with comma.

**Comment by eliphatfs** [comment]
> This is pretty simple actually, just use "with accelerator.accumulate(model1), accelerator.accumulate(model2): " this is the mechanism of "with", the following code will be in this two contexts, so just simply put them together with comma.

This apparently is not working. I printed AdamW statistics of parameter groups from different models, and one of them will go out of sync between GPUs with this setup, which from my point of view should not happen in DDP.

Quoting from `torch` forum:
> That said, if a single call to backward involves gradient accumulation for more than 1 DDP wrapped module, then you’ll have to use a different process group for each of them to avoid interference.

Format:
```
Sync [GPU ID]
Min Max Mean of parameter group exp_sq_avg.sqrt()
...
```

![image](https://github.com/huggingface/accelerate/assets/23738781/f95933f7-3074-48bb-b7f0-06aeb07a7ec8)

After wrapping the models in a SuperModel module it no longer goes async.

![image](https://github.com/huggingface/accelerate/assets/23738781/5a2e526b-bb8d-4f17-af28-5272d75ae892)

**Comment by eliphatfs** [comment]
> This is pretty simple actually, just use "with accelerator.accumulate(model1), accelerator.accumulate(model2): " this is the mechanism of "with", the following code will be in this two contexts, so just simply put them together with comma.

This apparently is not working. I printed AdamW statistics of parameter groups from different models, and one of them will go out of sync between GPUs with this setup, which from my point of view should not happen in DDP.

Quoting from `torch` forum:
> That said, if a single call to backward involves gradient accumulation for more than 1 DDP wrapped module, then you’ll have to use a different process group for each of them to avoid interference.

Format:
```
Sync [GPU ID]
Min Max Mean of parameter group exp_sq_avg.sqrt()
...
```

![image](https://github.com/huggingface/accelerate/assets/23738781/f95933f7-3074-48bb-b7f0-06aeb07a7ec8)

After wrapping the models in a SuperModel module it no longer goes async.

![image](https://github.com/huggingface/accelerate/assets/23738781/5a2e526b-bb8d-4f17-af28-5272d75ae892)

**Comment by eliphatfs** [comment]
TL; DR: don't do gradient accumulation with multiple models. Wrap them in a wrapper model and do accelerator stuff with it. Move relevant forward logic inside the wrapper model.

**Comment by mahdip72** [comment]
@eliphatfs would you please show your solution (wrapping models together) in a sudo code?

**Comment by eliphatfs** [comment]
Basically, if you have this in your main training loop:
```python
states = text_encoder(input_ids)
pred = unet(noisy_latents, states, timesteps)
loss = F.mse_loss(pred, targets)
# now loss.backward() will corrupt gradients if you are using accumulation on multi-gpu
```
Change it into:
```python
class SuperModel(nn.Module):
    def __init__(self, unet: UNet2DConditionModel, text_encoder: nn.Module) -> None:
        super().__init__()
        self.unet = unet
        self.text_encoder = text_encoder
  def forward(self, input_ids, noisy_latents, timesteps):
        states = text_encoder(input_ids)
        return unet(noisy_latents, states, timesteps)
```
When constructing models, construct a `SuperModel` after you do with the modules.
When `accelerator.prepare`, only do it on the SuperModel.
And in the main loop replace the two lines with a single call to `SuperModel` forward:
```python
pred = supermodel(noisy_latents, states, timesteps)
loss = F.mse_loss(pred, targets)
```
You may also need to change the final saving:
```python
        supermodel: SuperModel = accelerator.unwrap_model(supermodel)
        supermodel.text_encoder.save_pretrained(os.path.join(args.output_dir, 'text_encoder'))
        supermodel.unet.save_pretrained(os.path.join(args.output_dir, 'unet'))
```
I have not yet a good idea how to do with LoRA layers yet. It seems that LoRA layers on multiple modules are causing more problems since only `AttnProcLayers` get `prepare`-d.

**Comment by mahdip72** [comment]
> I "solved" it by creating one `Accelerator` per model. If you use only one and register the models via `accelerator.prepare(model1, ..., modelN)` at least one of the models is not learning anything. This might be a bug.

You mean create two accelerator objects and use nested accumulate for training loop?
``` python
with accel_1.accumulate(model1):
     with accel_2.accumulate(model2):
          training_steps
```

**Comment by cfeng16** [comment]
Does it now support gradient accumulation for multiple models?

**Comment by hkunzhe** [comment]
> Does it now support gradient accumulation for multiple models?

I think #1708 should fix it according to [comment](https://github.com/huggingface/diffusers/issues/4046#issuecomment-1686548737).

**Comment by cfeng16** [comment]
Can we use gradient accumulation for multiple models in distributed training?

**Comment by muellerzr** [comment]
Yes, just wrap them all in the accumulate function as shown earlier

**Comment by Chao0511** [comment]
> As I think, in this case writing accumulation by yourself maybe more flexible. `Accelerator.accumulate()` is not necessary. Just write code like:
> 
> ```python
> loss = loss / gradient_accumulation_steps
> accelerator.backward(loss)
> if (index+1) % gradient_accumulation_steps == 0:
>     optimizer.step()
>     scheduler.step()
>     optimizer.zero_grad()
> ```

don't forget to delete with accelerator.accumulate(unet): and Accelerator(gradient_accumulation_steps=args.gradient_accumulation_steps, guys
